hnn_core/gui/_viz_manager.py

```diff
@@ -0,0 +1,475 @@
+            display(widgets['figs_tabs'])
+
+    _prev_titles = copy.deepcopy(widgets['figs_tabs'].titles)
+    widgets['figs_tabs'].selected_index = None
+    widgets['figs_tabs'].children = widgets['figs_tabs'].children + (
+        fig_outputs, )
+
+    widgets['figs_tabs'].titles = _prev_titles + (_idx2figname(fig_idx), )
+
+    with fig_outputs:
+        dpi = _dpi
+        figsize = (scale * ((int(viz_output_layout.width[:-2]) - 10) / dpi),
+                   scale * ((int(viz_output_layout.height[:-2]) - 10) / dpi))
+        mosaic = fig_templates[template_name]['mosaic']
+        kwargs = eval(f"dict({fig_templates[template_name]['kwargs']})")
+        fig, axd = plt.subplot_mosaic(mosaic, figsize=figsize, dpi=dpi,
+                                      **kwargs)
+        fig.tight_layout()
+        fig.canvas.header_visible = False
+        fig.canvas.footer_visible = False
+        plt.show()
+
+    _add_axes_controls(widgets, data, fig=fig, axd=axd)
+
+    data['figs'][fig_idx] = fig
+    data['fig_idx']['idx'] += 1
+    widgets['figs_tabs'].selected_index = n_tabs
+
+
+class _VizManager:
+    """GUI visualization panel manager class.
+
+    Parameters
+    ----------
+    gui_data : dict
+        A dict containing all simulation data
+    viz_layout : dict
+        A dict about visualization layout specs
+    """
+    def __init__(self, gui_data, viz_layout):
+        plt.close("all")
+        self.viz_layout = viz_layout
+
+        self.axes_config_output = Output()
+        self.figs_output = Output(
+            layout=self.viz_layout['visualization_window'])
+
+        # widgets
+        self.axes_config_tabs = _TabWithNoneIndex()
+        self.figs_tabs = _TabWithNoneIndex()
+        self.axes_config_tabs.selected_index = None
+        self.figs_tabs.selected_index = None
+        link(
+            (self.axes_config_tabs, 'selected_index'),
+            (self.figs_tabs, 'selected_index'),
+        )
+
+        template_names = list(fig_templates.keys())
+        self.templates_dropdown = Dropdown(
+            description='Layout template:',
+            options=template_names,
+            value=template_names[0],
+            style={'description_width': 'initial'},
+            layout=Layout(width="98%"))
+        self.make_fig_button = Button(
+            description='Make figure',
+            button_style="primary",
+            style={'button_color': self.viz_layout['theme_color']},
+            layout=self.viz_layout['btn'])
+        self.make_fig_button.on_click(self.add_figure)
+
+        # data
+        self.fig_idx = {"idx": 1}
+        self.figs = {}
+        self.gui_data = gui_data
+
+    @property
+    def widgets(self):
+        return {
+            "figs_output": self.figs_output,
+            "axes_config_tabs": self.axes_config_tabs,
+            "figs_tabs": self.figs_tabs,
+            "templates_dropdown": self.templates_dropdown
+        }
+
+    @property
+    def data(self):
```

Comment:
Document as attributes? Since this is getting used in a separate file it'd be good to make a traceback easy